File: Python/Arrays & Hashing/contains-duplicate.py
# Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.

## Sol 1 - Hash Map

class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        countMap = {}
        for i in range(len(nums)):
            if nums[i] in countMap and countMap[nums[i]] > 0:
                return True
            countMap[nums[i]] = countMap.get(nums[i], 0) + 1
        return False

## Sol 2 - Hash Set

# class Solution:
#     def containsDuplicate(self, nums: List[int]) -> bool:
#         seen = set() 
#         for num in nums: 
#             if num in seen:
#                 return True
#             seen.add(num)
#         return False

## Sol 3 - Sort Array

# class Solution:
#     def containsDuplicate(self, nums: List[int]) -> bool:
#         nums.sort()
#         n = len(nums)
#         for i in range(n - 1):
#             if nums[i] == nums[i + 1]:
#                 return True
#         return False

File: C#/Arrays & Hashing/ContainsDuplicate.cs
// https://leetcode.com/problems/contains-duplicate/

// Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.

// Sol 1 - Hash Map, O (n) space time

public class Solution {
    public bool ContainsDuplicate(int[] nums) {
        Dictionary<int, int> seen = new Dictionary<int, int>();
        for (int i = 0; i < nums.Length; i ++) {
            if (seen.ContainsKey(nums[i])) {
                return true; 
            }
            seen[nums[i]] = 1;
        }
        return false;
    }
}

// Sol 2 - Hash set, O(n) space time 

public class Solution {
    public bool ContainsDuplicate(int[] nums) {
        HashSet<int> seen = new HashSet<int>();

        foreach (int num in nums) {
            if (seen.Contains(num)) {
                return true;
            }
            seen.Add(num);
        }
        return false;
    }
}

// Sol 3 - Array sort O(nlogn) time, O(1) space

public class Solution {
    public bool ContainsDuplicate(int[] nums) {
        Array.Sort(nums); 
        for (int i = 0; i < nums.Length - 1; i++) {
            if (nums[i] == nums[i + 1]) {
                return true;
            }
        }
        return false;
    }
}